File: client_lourd/src/app/services/luck-services/coin-flip.service.ts
/* eslint-disable no-restricted-imports */
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { ConfirmationDialogComponent } from '@app/components/general-elements/confirmation-dialog/confirmation-dialog.component';
import { ErrorDialogComponent } from '@app/components/general-elements/error-dialog/error-dialog.component';
import { WIDTH_SIZE } from '@app/constants/constants';
import { CoinFlipEvents, CoinFlipGameState } from '@app/constants/enum-class';
import { SocketClientService } from 'src/app/services/websocket-services/general/socket-client-manager.service';
import { AuthService } from '../auth/auth.service';

@Injectable({
    providedIn: 'root',
})
export class CoinFlipService {
    gameState: CoinFlipGameState = CoinFlipGameState.BettingPhase;
    selectedSide: string;
    winningSide: string;
    betAmount: number = 0;
    submitted: boolean = false;
    time: number = 0;
    name: string = '';
    playerList: {
        heads: { name: string; bet: number }[];
        tails: { name: string; bet: number }[];
    } = { heads: [], tails: [] };
    history: string[] = [];

    constructor(
        private socketService: SocketClientService,
        private matdialog: MatDialog,
        private authService: AuthService,
    ) {
        this.resetAttributes();
        this.initializeEventListeners();
        this.subscribeToName();
    }

    subscribeToName(): void {
        this.authService.user$.subscribe((user) => {
            if (user) {
                this.name = user.username;
            }
        });
    }

    selectSide(side: string) {
        this.selectedSide = side;
    }

    resetAttributes() {
        this.gameState = CoinFlipGameState.BettingPhase;
        this.selectedSide = '';
        this.winningSide = '';
        this.betAmount = 0;
        this.time = 0;
        this.submitted = false;
    }

    getState() {
        this.socketService.send(
            CoinFlipEvents.JoinGame,
            (answer: {
                playerList: {
                    heads: { name: string; bet: number }[];
                    tails: { name: string; bet: number }[];
                };
                history: string[];
                state: CoinFlipGameState;
            }) => {
                this.playerList = answer.playerList;
                this.history = answer.history;
                this.gameState = answer.state;
                const allPlayers = [...answer.playerList.heads, ...answer.playerList.tails];
                const isMePlaying = allPlayers.some((player) => player.name === this.name);

                if (!isMePlaying && this.gameState === CoinFlipGameState.BettingPhase) {
                    this.resetAttributes();
                }
            },
        );
    }

    removeEventListeners() {
        this.socketService.socket.off(CoinFlipEvents.StartGame);
        this.socketService.socket.off(CoinFlipEvents.PreFlippingPhase);
        this.socketService.socket.off(CoinFlipEvents.FlippingPhase);
        this.socketService.socket.off(CoinFlipEvents.Results);
        this.socketService.socket.off(CoinFlipEvents.SendPlayerList);
        this.socketService.socket.off(CoinFlipEvents.BetTimeCountdown);
    }

    initializeEventListeners() {
        this.removeEventListeners();
        this.socketService.on(CoinFlipEvents.StartGame, () => {
            this.resetAttributes();
        });

        this.socketService.on(CoinFlipEvents.PreFlippingPhase, () => {
            this.gameState = CoinFlipGameState.PreFlippingPhase;
        });

        this.socketService.on(CoinFlipEvents.FlippingPhase, () => {
            this.gameState = CoinFlipGameState.FlippingPhase;
        });

        this.socketService.on(
            CoinFlipEvents.Results,
            (answer: {
                result: string;
                playerList: {
                    heads: { name: string; bet: number }[];
                    tails: { name: string; bet: number }[];
                };
                history: string[];
            }) => {
                this.playerList = answer.playerList;
                this.history = answer.history;
                this.gameState = CoinFlipGameState.ResultsPhase;
                this.winningSide = answer.result;
            },
        );

        this.socketService.on(
            CoinFlipEvents.SendPlayerList,
            (playerList: { heads: { name: string; bet: number }[]; tails: { name: string; bet: number }[] }) => {
                this.playerList = playerList;
            },
        );

        this.socketService.on(CoinFlipEvents.BetTimeCountdown, (time: number) => {
            // turning seconds into ms. no need for const.
            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            this.time = time / 10;
        });
    }

    submitBet() {
        const dialogRef = this.matdialog.open(ConfirmationDialogComponent, {
            width: WIDTH_SIZE,
            panelClass: 'custom-container',
            data: `Voulez vous miser la somme de : ${this.betAmount} pièces?`,
        });

        dialogRef.afterClosed().subscribe((result) => {
            if (result) {
                if (this.betAmount > 0 && this.betAmount % 1 === 0 && this.gameState === CoinFlipGameState.BettingPhase) {
                    this.socketService.send(
                        CoinFlipEvents.SubmitChoice,
                        { choice: this.selectedSide, bet: this.betAmount },
                        (submitStatus: boolean) => {
                            if (submitStatus) {
                                this.submitted = true;
                            } else {
                                this.matdialog.open(ErrorDialogComponent, {
                                    width: WIDTH_SIZE,
                                    panelClass: 'custom-container',
                                    data: {
                                        message:
                                            'Vous ne pouvez pas parier plus de pièces que celles détenues ou bien parier en dehors de la phase de mise!',
                                        reloadOnClose: false,
                                    },
                                });
                            }
                        },
                    );
                } else if (this.betAmount === 0 || this.betAmount % 1 !== 0) {
                    this.matdialog.open(ErrorDialogComponent, {
                        width: WIDTH_SIZE,
                        panelClass: 'custom-container',
                        data: { message: 'Vous ne pouvez pas parier 0 pièces.', reloadOnClose: false },
                    });
                } else if (this.gameState !== CoinFlipGameState.BettingPhase) {
                    this.errorOutsideBettingPhase();
                }
            }
        });
    }

    errorOutsideBettingPhase() {
        this.matdialog.open(ErrorDialogComponent, {
            width: WIDTH_SIZE,
            panelClass: 'custom-container',
            data: { message: 'Vous ne pouvez pas parier en dehors de la phase de mise.', reloadOnClose: false },
        });
    }

    updateBetAmount(value: number) {
        if (value < 0 || typeof value !== 'number') {
            return Number(0);
        }
        if (value % 1 !== 0) {
            return Math.ceil(Number(value));
        }
        return Number(value);
    }

    increaseBet(amount: number) {
        if (!this.betAmount) {
            this.betAmount = 0;
        }
        this.betAmount += amount;
    }
}

File: client_leger/lib/business/coinflip_service.dart
import 'package:client_leger/UI/confirmation/confirmation_dialog.dart';
import 'package:client_leger/UI/error/error_dialog.dart';
import 'package:client_leger/backend-communication-services/socket/websocketmanager.dart';
import 'package:client_leger/utilities/socket_events.dart';
import 'package:flutter/material.dart';

class CoinflipService {
  static final CoinflipService _instance = CoinflipService._internal();

  factory CoinflipService() {
    return _instance;
  }

  CoinflipService._internal() {
    _webSocketManager = WebSocketManager.instance;
  }

  late WebSocketManager _webSocketManager;

  CoinFlipGameState gameState = CoinFlipGameState.BettingPhase;
  String selectedSide = '';
  String winningSide = '';
  int betAmount = 0;
  bool submitted = false;
  int time = 0;
  Map<String, List<Map<String, dynamic>>> playerList = {
    'heads': [],
    'tails': []
  };
  List<String> history = [];

  selectSide(String side) {
    selectedSide = side;
  }

  resetAttributes() {
    gameState = CoinFlipGameState.BettingPhase;
    selectedSide = '';
    winningSide = '';
    betAmount = 0;
    time = 0;
    submitted = false;
  }

  getState() {
    _webSocketManager.webSocketSender(
      CoinFlipEvents.JoinGame.value,
      null,
      (answer) => {
        playerList = {
          'heads':
              List<Map<String, dynamic>>.from(answer['playerList']['heads']),
          'tails':
              List<Map<String, dynamic>>.from(answer['playerList']['tails']),
        },
        history = List<String>.from(answer['history']),
        gameState = CoinFlipGameState.values.firstWhere(
          (e) => e.toString() == 'CoinFlipGameState.${answer['state']}',
        ),
      },
    );
  }

  initializeEventListeners() {
    _webSocketManager.webSocketReceiver(
        CoinFlipEvents.StartGame.value, (_) => {resetAttributes()});

    _webSocketManager.webSocketReceiver(
        CoinFlipEvents.PreFlippingPhase.value,
        (_) => {
              gameState = CoinFlipGameState.PreFlippingPhase,
            });

    _webSocketManager.webSocketReceiver(
        CoinFlipEvents.FlippingPhase.value,
        (_) => {
              gameState = CoinFlipGameState.FlippingPhase,
            });

    _webSocketManager.webSocketReceiver(
        CoinFlipEvents.Results.value,
        (answer) => {
              playerList = {
                'heads': List<Map<String, dynamic>>.from(
                    answer['playerList']['heads']),
                'tails': List<Map<String, dynamic>>.from(
                    answer['playerList']['tails']),
              },
              history = List<String>.from(answer['history']),
              gameState = CoinFlipGameState.ResultsPhase,
              winningSide = answer['result'],
            });

    _webSocketManager.webSocketReceiver(
        CoinFlipEvents.SendPlayerList.value,
        (playerList) => {
              playerList = {
                'heads': List<Map<String, dynamic>>.from(
                    playerList['playerList']['heads']),
                'tails': List<Map<String, dynamic>>.from(
                    playerList['playerList']['tails']),
              },
            });

    _webSocketManager.webSocketReceiver(
        CoinFlipEvents.BetTimeCountdown.value,
        (newTime) => {
              time = newTime / 10,
            });
  }

  submitBet(BuildContext context) {
    showConfirmationDialog(
        context,
        "Voulez vous miser la somme de : $betAmount coins?",
        () => onSubmitBetConfirm(context));
  }

  onSubmitBetConfirm(BuildContext context) {
    if (betAmount > 0 &&
        betAmount % 1 == 0 &&
        gameState == CoinFlipGameState.BettingPhase) {
      _webSocketManager.webSocketSender(
          CoinFlipEvents.SubmitChoice.value,
          {'choice': selectedSide, 'bet': betAmount},
          (submitStatus) => {
                if (submitStatus)
                  {
                    submitted = true,
                  }
                else
                  {
                    showErrorDialog(
                      context,
                      'Vous ne pouvez pas parier plus de coins que ceux détenus ou bien parier en dehors de la phase de mise!',
                    ),
                  },
              });
    } else if (betAmount == 0 || betAmount % 1 != 0) {
      showErrorDialog(context, 'Vous ne pouvez pas parier 0 coins.');
    } else if (gameState != CoinFlipGameState.BettingPhase) {
      errorOutsideBettingPhase(context);
    }
  }

  errorOutsideBettingPhase(BuildContext context) {
    showErrorDialog(
        context, 'Vous ne pouvez pas parier en dehors de la phase de mise.');
  }

  updateBetAmount(int value) {
    if (value < 0) {
      return 0;
    }
    if (value % 1 != 0) {
      return value.ceil();
    }
    return value;
  }

  increaseBet(int amount) {
    betAmount += amount;
  }
}
